ciao-controller/internal/datastore/datastore.go

```diff
@@ -1140,6 +1166,90 @@ func (ds *Datastore) DeleteInstance(instanceID string) error {
 	return nil
 }
 
+func (ds *Datastore) restartInstance(instanceID string) error {
+	ds.instancesLock.Lock()
+	i := ds.instances[instanceID]
+	i.State = payloads.Pending
+	ds.instancesLock.Unlock()
+
+	instanceStat := types.CiaoServerStats{
+		ID:        instanceID,
+		Timestamp: time.Now(),
+		Status:    payloads.Pending,
+	}
+	ds.instanceLastStatLock.Lock()
+	ds.instanceLastStat[instanceID] = instanceStat
+	ds.instanceLastStatLock.Unlock()
+
+	stats := []payloads.InstanceStat{
+		{
+			InstanceUUID: instanceID,
+			State:        payloads.Pending,
+		},
+	}
+
+	return errors.Wrapf(ds.db.addInstanceStats(stats, ""), "error adding instance stats to database")
+}
+
+// RestartInstance resets a restarting instance's state to pending.
+func (ds *Datastore) RestartInstance(instanceID string) error {
+	err := ds.restartInstance(instanceID)
+	if err != nil {
+		return errors.Wrapf(err, "error restarting instance")
+	}
+	return nil
+}
+
+func (ds *Datastore) stopInstance(instanceID string) error {
+	ds.instancesLock.Lock()
+	i := ds.instances[instanceID]
+	oldNodeID := i.NodeID
+	i.NodeID = ""
+	i.State = payloads.Exited
+	ds.instancesLock.Unlock()
+
+	ds.tenantsLock.Lock()
+	tenant := ds.tenants[i.TenantID]
+	if tenant != nil {
+		tenant.removeInstanceResources(i)
+	}
+	ds.tenantsLock.Unlock()
+
+	// we may not have received any node stats for this instance
+	if oldNodeID != "" {
+		ds.nodesLock.Lock()
+		delete(ds.nodes[oldNodeID].instances, instanceID)
+		ds.nodesLock.Unlock()
+	}
+
+	instanceStat := types.CiaoServerStats{
```

Comment:
I should probably factorize this code as there is some very similiar code in the preceeding function.